From afd0e19f1d198b3807d137be40cb845a28ccd2d5 Mon Sep 17 00:00:00 2001
From: Tasos Sahanidis <tasos@tasossah.com>
Date: Sun, 15 Mar 2020 22:53:13 +0200
Subject: [PATCH] Enable only mouse buttons on the mouse device

This is needed as a workaround to prevent libgdx games from crashing
whenever a mouse button is pressed
---
 src/daemon/input_linux.c | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/src/daemon/input_linux.c b/src/daemon/input_linux.c
index 0399212..8c8a604 100644
--- a/src/daemon/input_linux.c
+++ b/src/daemon/input_linux.c
@@ -16,16 +16,21 @@ int uinputopen(struct uinput_user_dev* indev, int mouse){
             return 0;
         }
     }
-    // Enable all keys and mouse buttons
+
     ioctl(fd, UI_SET_EVBIT, EV_KEY);
-    for(int i = 0; i < KEY_CNT; i++)
-        ioctl(fd, UI_SET_KEYBIT, i);
     if(mouse){
+        // Enable only the mouse buttons as a workaround for libgdx crashing
+        // https://github.com/libgdx/libgdx/issues/5857
+        for(int i = BTN_LEFT; i <= BTN_TASK; i++)
+            ioctl(fd, UI_SET_KEYBIT, i);
         // Enable mouse axes
         ioctl(fd, UI_SET_EVBIT, EV_REL);
         for(int i = 0; i < REL_CNT; i++)
             ioctl(fd, UI_SET_RELBIT, i);
     } else {
+        // Enable all possible keys
+        for(int i = 0; i < KEY_CNT; i++)
+            ioctl(fd, UI_SET_KEYBIT, i);
         // Enable LEDs
         ioctl(fd, UI_SET_EVBIT, EV_LED);
         //for(int i = 0; i < LED_CNT; i++)
-- 
2.26.2

